### ISSUE TITLE
Deprecation warning regarding Gradle 5.0 when using plugin

### ISSUE THREAD

**Issue by picaron** [bmuschko/gradle-docker-plugin#725]
### Expected Behavior

The plugin should continue working when migrating to Gradle 6.0

### Current Behavior

Currently when using Gradle 5.0, we get the following deprecation warning:

```
The DefaultSourceDirectorySet constructor has been deprecated. This is scheduled to be removed in Gradle 6.0. Please use the ObjectFactory service to create instances of SourceDirectorySet instead.
The ProjectLayout.directoryProperty() method has been deprecated. This is scheduled to be removed in Gradle 6.0. Please use the ObjectFactory.directoryProperty() method instead.
        at build_9t0xk9xnqi0dg4yogioioq88k.run(/Users/pascal.chouinard/projects/product-management2/build.gradle:32)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
The DefaultTask.newOutputFile() method has been deprecated. This is scheduled to be removed in Gradle 6.0. Please use the ObjectFactory.fileProperty() method instead.
        at build_9t0xk9xnqi0dg4yogioioq88k.run(/Users/pascal.chouinard/projects/product-management2/build.gradle:207)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
The DefaultTask.newInputDirectory() method has been deprecated. This is scheduled to be removed in Gradle 6.0. Please use the ObjectFactory.directoryProperty() method instead.
        at build_9t0xk9xnqi0dg4yogioioq88k.run(/Users/pascal.chouinard/projects/product-management2/build.gradle:221)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
The DefaultTask.newInputFile() method has been deprecated. This is scheduled to be removed in Gradle 6.0. Please use the ObjectFactory.fileProperty() method instead.
        at build_9t0xk9xnqi0dg4yogioioq88k.run(/Users/pascal.chouinard/projects/product-management2/build.gradle:221)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
```

### Context

I want my builds to keep working when upgrading to Gradle 6.0

### Steps to Reproduce (for bugs)

Use the gradle wrapper 5.0 and the gradle docker plugin in a project to see the warning

### Your Environment

Using gradle wrapper 5.0

**Comment by bmuschko** [comment]
Many of the replacement methods have been introduced with Gradle 5.0 e.g. [`ObjectFactory.directoryProperty()`](https://docs.gradle.org/current/javadoc/org/gradle/api/model/ObjectFactory.html#directoryProperty--). The plugin, however, is still supposed to be compatible with its minimum supported version Gradle 4.8.

So if we want to support Gradle 4.x then we'll still need to keep on using the "old" methods. I don't think we'll change that until we make Gradle 5.0 the official minimum version for this plugin. Keep in mind that Gradle 5.0 just came out 22(!) days ago so I don't think it would be fair to users to require all of their project to switch over right away. It will definitely take another couple of months before we'll make that switch.

**Comment by cdancy** [comment]
@bmuschko I mean ... we've made all of these changes already ;) We're still a ways away from catching up here internally.

**Comment by bmuschko** [comment]
> I mean ... we've made all of these changes already.

What changes specifically? Are you referring to changes in this code base?

**Comment by cdancy** [comment]
It was more a tongue in cheek joke about all the changes you've made recently.

But to your comment ... We in theory could , through groovy magic, write things in a way that would work across the board whether or not the necessary methods are present. Whether it would be pretty or not is another thing but it could work.

**Comment by picaron** [comment]
I agree it may be a bit too early at this stage if this will break compatibility for gradle 4.x users

**Comment by denuno** [comment]
There are a lot of Gradle plugins in a similar boat.  It's going to be rough for a bit here with Java > 1.8 and Gradle > 4, etc. (IDEs and whatnot are all hauling ass to get up to speed while staying compatible as well) just in general.

I couldn't think of any way to handle the transition besides the time-honored tradition of tossing in some shims in the meantime.  Forking and backporting are options as well, I guess, but yuck (in general-- it can be a more attractive option, in some cases (performance-related mostly)).

FWIW Groovy meta-programming seems like the most viable approach for Gradle shims.  Everything is ugly from some perspective...  I reckon it's a pick-your-poison type situation.

**Comment by bmuschko** [comment]
I'd like to avoid any backward-compatibility hacks if possible especially if we are dealing with just minor issues like warning messages. It just makes maintaining the code extremely hard plus you have to write tests against multiple versions of Gradle to ensure that it works as intended. I'd rather just switch over at some point of time and make 5.0 the minimum version.

**Comment by denuno** [comment]
I was curious as to if I was talking out my ass about how hard getting rid of the error would be.

It turns out it can be just a few lines of code, if you limit the scope-- or a bunch of lines if you want to be fancy about it.

The latest Kotlin plugin addressed a similar issue via: https://github.com/JetBrains/kotlin/commit/67e82a54e5ee529116e881953f93a4c8f216e33a

Interestingly it looks to be checking both for the Gradle version and for the deprecation annotation.  Also interesting is the Gradle source for what they're checking.  According to a code comment it doesn't appear to _really_ be going away in v6 -- they're mostly wanting people to switch to the object route and to drop an overloaded method -- so I think maybe the deprecation warning might even go away on its own depending on how you're using it (neither here nor there, but an interesting thing to think about if you're checking both, and regarding the general idea of backwards compatible maintenance... looks like Kotlin is supporting Gradle all the way back to < 2.12 without _too_ much trouble, relatively speaking.).

I need to address this warning myself in another plugin, so I'll probably add a functional test for whatever it turns out the lowest Gradle version is I can support.  If I ever get around to doing that I'll see if I can't do a PR for this plugin to test the minimum supported Gradle version too, as I didn't see anything when I took a quick peek, and manually testing various versions of stuff *is* a PITA.

It's just amazing how there are bits of programming that apply to any language-- reminds me of the awe I felt when learning to read sheet music: "I can play anything now, muahahaha!" (not really how it works, but kinda).  Apologies for the issue spam, I thought it was neat information, and maybe helpful conceptually to somebody somewhere.

TL;DR: I :heart: programming and open source software.

**Comment by bmuschko** [comment]
> It's all relative. One person's minor warning message issue is another person's multiple screens of noise annoyance.

I totally understand that deprecation or warning messages can be bothersome. Optimally, those should be fixable. Your answer mixes up some of the concerns though. Let me try to clarify in the following paragraphs.

> There is also a time-honored tradition of telling folk to ignore ignorable warnings, and yet still having some panic when they see them.

Gradle doesn't render deprecation messages by default - only if you provide a command line option. It only says that there _are_ deprecation messages. So you'd really only see a list of message if you enable them, otherwise they are hidden. I'd say the pain is somewhat small for the "regular" user.

> I found the Groovy reflection warnings when running on Java 1.8+ annoying mostly because hiding them also hides unrelated warnings.

This message is not a deprecation warning and didn't originate from the Gradle code base. It still is or at least was an issue in the Groovy library. To ensure backward compatibility with older Java versions, Gradle couldn't upgrade to a newer version of Groovy (which fixed the issue). I believe the Gradle team pulled some tricks to avoid rendering the message. So overall this is unrelated to Gradle API deprecation messages.

> It turns out it can be just a few lines of code, if you limit the scope-- or a bunch of lines if you want to be fancy about it.

I am pretty sure it will be more work than you think for little gain. It will have to be fixed in multiple places. Optimally, you should also run all functional tests with at least the current version of Gradle and the minimum compatible version. Additionally, it will have side effects users won't see like additional dependencies for testing, an increased execution time for running all tests, more code to run the tests (e.g. a JUnit Runner implementation), setting up CI to run those suites in parallel etc. 

To be honest, I am personally not willing to spend time on it. Maybe @cdancy wants to work on it as he is the main committerof this plugin. In say 3 months we can release a version that will require Gradle 5.0 and avoid implementing any hacks in the code base.

**Comment by denuno** [comment]
For CI I like to set the warning-mode to `all` in gradle.properties, and I also enable stack traces (mostly because I'm lazy and/or might not have the same setup for whatever reason locally, and those settings make it easy to see what it's actually picking up), so there's a lot more output than a "normal" run would have.  As a result, I don't normally have that summary line, so it's something new and thus potentially scary. :ghost:

I know that Groovy warning doesn't come from Gradle, it was just an example of the phenomenon.  I copied the same things they used to get rid of the messages in my builds back when they added that, just to silence them until the new Groovy is in Gradle.  I'm just not a fan of silencing expected warnings when it has the side-effect of silencing the non-expected ones as well.  (E.g. I could grab system.out, or something fun like that, to silence everything... but...)

When I was experimenting over the weekend, I added a GradleShim class with I think 4 methods, which handled all the deprecated warnings I was getting (except for the one from Kotlin itself).  If @cdancy is interested I could do a PR with that class, but there's a good argument for not being backwards compatible and pushing folks to get on 5+.  Heck, 5.1 is just around the corner!  It's kind of part of the new mindset of "release early, release often", which has spread across the software landscape in the last few years.

Testing multiple versions does add overhead.  I found https://plugins.gradle.org/plugin/org.ysb33r.gradletest when I took a look, which made me happy.  It will add a good bit of time, so I wouldn't make it part of the standard test runs, but not a bad idea for CI.  Not that it's hard to manually test minimum and maximums, but when you have a lot of projects, that "requires a human" time does kind of add up.  I'd rather just set up a job to do it for me automatically.  Setting up a job is relatively quick and it only takes a few runs before it has paid for itself, time-wise.

Anyhow, despite how I might sound, I'm not advocating or asking for anyone to do anything.  Maybe advocating a little, but it's more just general thoughts, that I'm not sure are "correct" even.  It's quite likely that what I'm saying isn't too hard, is still a Bad Idea in the long run, and actually slows progress versus speeding it up.  Seems like it all depends on what the target audience is, their constraints, what the goals of the project are, etc. etc..  Plus, the times, they are a changing.  I think I read that Google is deprecating desktop Chromecast setup for Mac and PC with the next version of Chrome.  That's kind of amazing.  But I digress. Heh.  Happy New Year, and thanks for taking the time to read this stuff!

**Comment by bmuschko** [comment]
I realized that the plugin already uses a Gradle 5.0 method. Happy to accept a PR that fixes the deprecation message.

**Comment by cdancy** [comment]
lets close this for now as there is nothing immediate that needs to be done on our end and as Ben said: if folks want to submit a PR they are free to do so and we will certainly look it over.
